### PR TITLE
DOCSP-13859: UpdateOne usage example

### PR DESCRIPTION
**Pull request by rustagir**
### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-13859

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/golang/docsworker-xlarge/DOCSP-13859-update-one-usage-ex/usage-examples/updateOne/

**TODO:** edit links to Fundamentals pages "Change a Document" and "Retrieve Data" when they are created

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

### CODE REVIEW

File: source/usage-examples/updateOne.txt [hunk 1]
```diff
@@ -3,3 +3,68 @@ Update a Document
 =================
 
 .. default-domain:: mongodb
+
+You can update a document in a collection by using the ``Collection.UpdateOne()``
```
Comment (terakilobyte) [inline source/usage-examples/updateOne.txt:7]: suggestion: I think it's ok to shorten this to just ``UpdateOne()``.

```suggestion
You can update a document in a collection by using the ``UpdateOne()``
```

File: source/usage-examples/updateOne.txt [hunk 2]
```diff
@@ -3,3 +3,68 @@ Update a Document
...
+method. 
+
+This example passes a query filter and an update parameter to the
+``UpdateOne()`` method. The query filter specifies the document to
+match, while the update parameter specifies the change to
+make to the document. The following code example calls the
+``UpdateOne()`` method on a document matched by its ``_id`` field from
+the ``restaurants`` collection to create a new field called
+``avg_rating`` with a value of 4.4:
```
Comment (terakilobyte) [inline source/usage-examples/updateOne.txt:16]: nit/suggestion: Since we're going for streamlined and simple, I'd put slightly less emphasis on the query filter. Additionally, I don't think it's necessary to say `the following code example`, just `the following example`.

```suggestion
The following example passes a query filter and an update parameter to the
``UpdateOne()`` method, matching a document in the ``restaurants`` collection
by the value of its ``_id`` field and creating a new field in that document called
``avg_rating`` with a value of *4.4*:
```

File: source/usage-examples/updateOne.txt [hunk 3]
```diff
@@ -3,3 +3,68 @@ Update a Document
...
+   :language: go
+   :dedent:
+
+Click here <TODO: github link to file> to see a fully runnable example.
+
+.. note::
+
+   If the query filter does not match any documents, the operation will
+   return a result with a ``MatchedCount`` property of 0. If the filter
+   matches multiple documents, the method will select and update one document.
```
Comment (terakilobyte) [inline source/usage-examples/updateOne.txt:33]: question/comment: Shouldn't this be in the Fundamental page?

File: source/usage-examples/updateOne.txt [hunk 4]
```diff
@@ -3,3 +3,68 @@ Update a Document
...
+
+Expected Result
+---------------
+
+If you execute a :doc:`find
+</fundamentals/crud/read-operations/retrieve/#find>` operation querying
```
Comment (terakilobyte) [inline source/usage-examples/updateOne.txt:39]: suggestion: Let's mark this link as todo so we don't forget later.

```suggestion
If you execute a <TODO:`find
 /fundamentals/crud/read-operations/retrieve/#find>` operation querying
```

File: source/usage-examples/updateOne.txt [hunk 5]
```diff
@@ -3,3 +3,68 @@ Update a Document
...
+Additional Information
+----------------------
+
+For more information on updating documents, specifying query filters, and
+handling potential errors, see our guide on :doc:`changing a document
+</fundamentals/crud/write-operations/change-a-document/>`.
+
+For more information on update operators,
+see the :manual:`MongoDB update operator reference documentation
+</reference/operator/update/#update-operators>`.
```
Comment (terakilobyte) [inline source/usage-examples/updateOne.txt:64]: Same as the previous suggestion, let's clearly mark links that don't work as TODO

File: source/usage-examples/updateOne.txt [hunk 6]
```diff
@@ -3,3 +3,68 @@ Update a Document
...
+
+API Documentation
+~~~~~~~~~~~~~~~~~
+
+- :go-api:`UpdateOne() <mongo#Collection.UpdateOne>`
+- :go-api:`UpdateResult <mongo#UpdateResult>`
```
Comment (terakilobyte) [inline source/usage-examples/updateOne.txt:70]: Relating to my feedback on removing the note, if that goes the API link for `UpdateResult` should go as well.

File: source/usage-examples/updateOne.txt [hunk 7]
```diff
@@ -3,3 +3,68 @@ Update a Document
...
+   return a result with a ``MatchedCount`` property of 0. If the filter
+   matches multiple documents, the method will select and update one document.
+
+Expected Result
+---------------
+
+If you execute a :doc:`find
+</fundamentals/crud/read-operations/retrieve/#find>` operation querying
+for the same ``_id`` value used in the example, you should see that the
+matched document now includes the ``avg_rating`` field:
```
Comment (terakilobyte) [inline source/usage-examples/updateOne.txt:41]: nit: Let's clarify this language a bit that this is the result of the operation.

> After you run the example, the document will be updated to the following values:

Or something more eloquent than that.

**Review by terakilobyte — CHANGES_REQUESTED** [review]
Initial feedback

File: source/usage-examples/updateOne.txt (hunk 1 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:7]: accepted

File: source/usage-examples/updateOne.txt (hunk 2 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:16]: accepted

File: source/usage-examples/updateOne.txt (hunk 3 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:33]: accepted

File: source/usage-examples/updateOne.txt (hunk 4 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:39]: accepted

File: source/usage-examples/updateOne.txt (hunk 6 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:70]: accepted

File: source/usage-examples/updateOne.txt (hunk 5 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:64]: done

File: source/usage-examples/updateOne.txt (hunk 7 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:41]: noted, changed the phrasing

**Review by terakilobyte — APPROVED** [review]
lgtm

**Comment by rustagir** [comment]
Hi @kevinAlbs, this PR is ready for you to review!

File: source/usage-examples/updateOne.txt [hunk 8]
```diff
@@ -3,3 +3,56 @@ Update a Document
...
+fundamental page>.
+
+For more information on update operators,
+see the :manual:`MongoDB update operator reference documentation
+</reference/operator/update/#update-operators>`.
+
+API Documentation
+~~~~~~~~~~~~~~~~~
+
+:go-api:`UpdateOne() <mongo#Collection.UpdateOne>`
```
Comment (matthewdale) [inline source/usage-examples/updateOne.txt:58]: This link in on the staging server goes to pkg.go.dev documentation for the `github.com/mongodb/mongo-go-driver` package, but the canonical package is `go.mongodb.org/mongo-driver`. They're technically the same, but it could be confusing if the reader doesn't realize that.

Please update to link to the docs at https://pkg.go.dev/go.mongodb.org/mongo-driver instead.

File: source/code-snippets/updateOne.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+func main() {
+	// Replace the uri string with your MongoDB deployment's connection string.
+	uri := "mongodb+srv://<username>:<password>@<cluster-address>/test?w=majority"
+
+	ctx := context.TODO()
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(uri))
```
Comment (matthewdale) [inline source/code-snippets/updateOne.go:19]: Using the same Context for every call in a program isn't a common use case and may be confusing to readers who haven't used the Go driver before. Please update each call to use a separate `context.TODO()`

File: source/usage-examples/updateOne.txt [hunk 10]
```diff
@@ -3,3 +3,56 @@ Update a Document
...
+.. include:: /includes/run-example-tip.rst
+
+.. literalinclude:: /code-snippets/updateOne.go
+   :start-after: begin updateone
+   :end-before: end updateone
+   :emphasize-lines: 6
+   :language: go
+   :dedent:
+
+Click here <TODO: github link to file> to see a fully runnable example.
```
Comment (matthewdale) [inline source/usage-examples/updateOne.txt:24]: Should this `TODO` be resolved before merging?

File: source/code-snippets/updateOne.go (hunk 9 above)
Comment (rustagir) [reply source/code-snippets/updateOne.go:19]: accepted

File: source/usage-examples/updateOne.txt (hunk 10 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:24]: No, we will revise the links before publishing!

File: source/usage-examples/updateOne.txt (hunk 8 above)
Comment (rustagir) [reply source/usage-examples/updateOne.txt:58]: hard coding the URL instead

File: source/code-snippets/updateOne.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+	myCollection := client.Database("sample_restaurants").Collection("restaurants")
+	id, _ := primitive.ObjectIDFromHex("5eb3d668b31de5d588f42a7a")
+	filter := bson.D{{"_id", id}}
+	update := bson.D{{"$set", bson.D{{"avg_rating", 4.4}}}}
+
+	result, err := myCollection.UpdateOne(context.TODO(), filter, update)
+	// end updateone
+
+	if err != nil {
+		log.Fatal(err)
```
Comment (matthewdale) [inline source/code-snippets/updateOne.go:39]: nit: The other two error handling blocks use `panic(err)` instead. Consider always using the same exit call when handling errors (preferably `panic(err)` because it's simple).

**Review by matthewdale — APPROVED** [review]
👍

File: source/usage-examples/updateOne.txt (hunk 8 above)
Comment (matthewdale) [reply source/usage-examples/updateOne.txt]: I just noticed the code documentation URL contains the `@v1.7.0` version. Is that link going to be in the published version? If so, consider removing the `@v1.7.0` so readers will be linked to the latest version.

Comment (rustagir) [reply source/usage-examples/updateOne.txt]: since our Docs are versioned, we want to link specifically to the 1.7.x api docs!

File: source/code-snippets/updateOne.go (hunk 11 above)
Comment (rustagir) [reply source/code-snippets/updateOne.go:39]: accepted

Comment (rustagir) [reply source/code-snippets/updateOne.go]: accepted

File: source/code-snippets/updateOne.go [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+	update := bson.D{{"$set", bson.D{{"avg_rating", 4.4}}}}
+
+	result, err := myCollection.UpdateOne(context.TODO(), filter, update)
+	// end updateone
+
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Printf("%v document was updated.\n", result.ModifiedCount)
```
Comment (kevinAlbs) [inline source/code-snippets/updateOne.go:41]: Nit: if no document matches, this reads as "0 document was updated". Perhaps reword this as `"Documents updated: %v\n"`?

**Review by kevinAlbs — COMMENTED** [review]
Looks great, I only have nits.

File: source/code-snippets/updateOne.go [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+func main() {
+	// Replace the uri string with your MongoDB deployment's connection string.
+	uri := "mongodb+srv://<username>:<password>@<cluster-address>/test?w=majority"
```
[the remark this answers is not in the document]
Comment (rustagir) [reply source/code-snippets/updateOne.go:15]: accepted

File: source/code-snippets/updateOne.go (hunk 12 above)
Comment (rustagir) [reply source/code-snippets/updateOne.go:41]: accepted

**Review by benjirewis — APPROVED** [review]

**Review by kevinAlbs — APPROVED** [review]
